### PR TITLE
Allow type arguments in generic tagged templates

### LINKED ISSUE #11947: Support passing generics to tagged template string
I was trying to write some type definition for the [styled-components](https://github.com/styled-components/styled-components) project, but it looks like there is no way to specify a generic when using a tagged template. So... here is a subset of the typedef I'm writing:

``` ts
declare module 'styled-components' {
    import * as React from "react";

    export function input<P>(values: TemplateStringsArray): React.StatelessComponent<P>;
}
```

This is how I'd like to use it:

``` ts
import * as React from 'react';

/// <reference path="../../../typings/styled-components/styled-components.d.ts" />

import styled from 'styled-components';

interface Props {
  placeholder: string;
}

const Input = styled.input<Props>`
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
`;

```

I know that it can also be possible to override Input.propTypes manually[1], but I was wondering if this could be an useful addition to Typescript.

**TypeScript Version:**  2.0.3

[1] Or even write a custom component and just use the `styled(Component)`style`` syntax

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This pull request allows users to pass generic type arguments to tagged template strings.

```ts
declare function styledComponent<Props>(strs: TemplateStringsArray): Component<Props>;

interface MyProps {
  name: string;
  age: number;
}

styledComponent<MyProps> `
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
`;

declare function tag<T>(strs: TemplateStringsArray, ...args: T[]): T;

// inference fails because 'number' and 'string' are both candidates that conflict
let a = tag<string | number> `${100} ${"hello"}`;
```

Fixes #11947

# Background

Tagged templates are a form of invocation introduced in ECMAScript 2015. Like call expressions, generic functions may be used in a tagged template and TypeScript will infer the type arguments utilized:

```ts
declare function tag<T>(strs: TemplateStringsArray, ...args: T[]): T;

let a = tag(100, 200);     // has type 'number'
let b = tag `Hello world`; // has type TemplateStringsArray
```

However, in some cases there are no inference candidates

```ts
declare function styledComponent<Props>(strs: TemplateStringsArray): Component<Props>

// has type 'Component<{}>' because there were no candidates
let a = styledComponent `
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
`;
```

or, type arguments cannot be inferred because TypeScript is conservative in its inferences

```ts
declare function tag<T>(strs: TemplateStringsArray, ...args: T[]): T;

// inference fails because 'number' and 'string' are both candidates that conflict
let b = tag `${100} ${"hello"}`;
```

# Parser changes

The core change is a new intermediate grammar production between *MemberExpression* and *CallExpression*/*NewExpression*

> *TaggedTemplateWithTypeArguments*:
> &nbsp;&nbsp;&nbsp;&nbsp;*TaggedTemplateWithTypeArguments* `<` TypeArgumentList `>` *TemplateLiteral*
> &nbsp;&nbsp;&nbsp;&nbsp;*MemberExpression*

Then *CoverCallExpressionAndAsyncArrowHead* (which is the grammar production that currently transitions a *CallExpression* into a *MemberExpression*) no longer references *MemberExpression* *Arguments*, but instead looks more like:

> *CoverCallExpressionAndAsyncArrowHead*:
> &nbsp;&nbsp;&nbsp;&nbsp;*TaggedTemplateWithTypeArguments* *Arguments*

Similarly, *NewExpression* no longer references `MemberExpression`, and instead goes for

> *NewExpression*:
> &nbsp;&nbsp;&nbsp;&nbsp;*TaggedTemplateWithTypeArguments* *TypeArgumentsArguments*
> &nbsp;&nbsp;&nbsp;&nbsp;new *TaggedTemplateWithTypeArguments* *Arguments*

Within our mechanics for *MemberExpression* still use the following production for tagged template expressions (which you can see in `parseMemberExpressionRest`)

> *MemberExpression*:
> &nbsp;&nbsp;&nbsp;&nbsp;*MemberExpression* *TemplateLiteral*

This way we always try to parse out a template after a *MemberExpression*, even when there are no type arguments.

### CODE REVIEW

**Comment by Havret** [comment]
Thank you so much! It really made my day. :)

**Comment by MartinJohns** [comment]
Can you please also update the spec according to the new changes?

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 1]
```diff
@@ -3516,8 +3517,8 @@ declare namespace ts {
     function updateCall(node: CallExpression, expression: Expression, typeArguments: ReadonlyArray<TypeNode> | undefined, argumentsArray: ReadonlyArray<Expression>): CallExpression;
     function createNew(expression: Expression, typeArguments: ReadonlyArray<TypeNode> | undefined, argumentsArray: ReadonlyArray<Expression> | undefined): NewExpression;
     function updateNew(node: NewExpression, expression: Expression, typeArguments: ReadonlyArray<TypeNode> | undefined, argumentsArray: ReadonlyArray<Expression> | undefined): NewExpression;
-    function createTaggedTemplate(tag: Expression, template: TemplateLiteral): TaggedTemplateExpression;
-    function updateTaggedTemplate(node: TaggedTemplateExpression, tag: Expression, template: TemplateLiteral): TaggedTemplateExpression;
+    function createTaggedTemplate(tag: Expression, typeArguments: NodeArray<TypeNode>, template: TemplateLiteral): TaggedTemplateExpression;
```
Comment (mhegazy) [inline tests/baselines/reference/api/tsserverlibrary.d.ts]: this is an API breaking change. we need to document it

Comment (mhegazy) [reply tests/baselines/reference/api/tsserverlibrary.d.ts]: the other option is to put it at the end, or have multiple overloads. check with @rbuckton

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -1032,17 +1032,19 @@ namespace ts {
             : node;
     }
 
-    export function createTaggedTemplate(tag: Expression, template: TemplateLiteral) {
+    export function createTaggedTemplate(tag: Expression, typeArguments: NodeArray<TypeNode>, template: TemplateLiteral) {
```
Comment (ajafff) [inline src/compiler/factory.ts]: the API would be easier to use if the parameter is typed as `ReadonlyArray` and it's converted to a `NodeArray` in the function body.
this also applies to the update function below.

Comment (DanielRosenwasser) [reply src/compiler/factory.ts]: Good call!

**Comment by MartinJohns** [comment]
No update of the spec. :-(

**Comment by sergeysova** [comment]
What about passed functions?

```ts
const Example = styled.div`
  font-size: ${p => p.size}rem;
  flex-direction: ${p => p.row ? 'row' : 'column'};
`
```

**Comment by DanielRosenwasser** [comment]
@MartinJohns it's on my backlog, sorry. 😞 

@sergey-shandar Not sure what you're asking about, but any problems or authoring questions should be filed as a new issue or as a StackOverflow question respectively.

**Comment by sergey-shandar** [comment]
@DanielRosenwasser I think you mean @sergeysova.

**Pull request merged**